### ISSUE TITLE
Channel Per Octave and MPE mode not properly exclusive

### ISSUE THREAD

**Issue by nylonius** [surge-synthesizer/surge#5908]
The mac version of Surge XT 1.1. does not respond to polyphonic aftertouch when MPE mode is active. Also if the Tune mode uses several midi channels (with octave shifts), Poly Aftertouch won't work.  Surge 1.9. has the same issue in MPE mode.

The latest nightly install of SurgeXT resulted in a weird interface with knobs missing in the faders, and items greyed out. But it seems to work - except for the missing aftertouch in MPE mode.

Polyphonic aftertouch does work if MPE mode is off, also in the Tune mode (provided the tuning uses one channel only).

MPE tuning works correctly. It's just the poly aftertouch which won't.
 
macBook Pro, Catalina; host Cubase. XT also tested as standalone.

**Comment by baconpaul** [comment]
What controller do you use? 

Mpe mode will listen to channel agtertouch for the voices and only listen to poly at on the global channel. If you want to modulate with, say, post key press pressure on a linnstrument that will appear as channel at per the spec 

this is odd I know but mpe is basicalky a set of mono synths per channel so doesn’t support voice level poly at

**Comment by baconpaul** [comment]
Oh and I fixed a skin bug last night in classic skin which may cause the things you saw with missing items

**Comment by baconpaul** [comment]
Oh as for poly at in tuning mode I bet that’s a real bug - if you have a device which can generate a simple midi file and a patch that doesn’t work attaching it here as a zip would be v helpful

**Comment by baconpaul** [comment]
Finally looking at the code I think I see why maybe poly at in tune octave mode doesn’t work - we look up based on key but have shifted the key already is my guess. Again a midi stream which is broken would help. Also adding @cgibbard

**Comment by nylonius** [comment]
Thanks. I'll see if i can make a midi file to demonstrate the problems. I'm using Lumatone as the controller, and H-pi UTE to tune the notes with dynamic multichannel pitchbend (MPE). This setup works with correctly tuned notes and polyphonic aftertouch in other MPE-friendly synths, such as Arturia Pigments, Pianoteq, u-he Diva and u-he Repro 5. So the problem seems to be in Surge. However, another Lumatone user has reported that in his setup, poly aftertouch does work in MPE mode; but he's not using a mac or Cubase as host.

**Comment by baconpaul** [comment]
Which version of surge are you using?

We fixed a problem with poly aftertouch about 5 days ago with multiple channels, after a lumatone user reported a problem (but that was't in MPE mode). I was able to test that in logic by faking the data too and it seemed to work.

But best would be

1. A MID file which shows the problem
2. A Surge FXP and
3. The exact version (copied from the about screen) of the plugin you are using
4. Which ideally would be the latest nightly

and I'm sure we can figure it out.

**Comment by nylonius** [comment]
Here's a multichannel midi file with poly aftertouch that works in other MPE synths but not Surge which doesn't respond to the aftertouch messages at all. The tuning data (pitch bends for each note) is not here; this is the MIDI that is sent to UTE for tuning and from there to the MPE synth. Cubase has a known bug that prevents a multichannel file from being recorded correctly when the MIDI is sent from external software, so I recorded the MIDI that's sent from Lumatone to UTE. If this is passed on from UTE to Surge or the other MPE synths, the tuning will be correct but Surge doesn't play the aftertouch. This is not a porblem in the host because it's also in the standalone version. I was using Surge XT 1.1.nnightly.455ecc4 
[Diva 01.mid.zip](https://github.com/surge-synthesizer/surge/files/8134707/Diva.01.mid.zip)
.

**Comment by baconpaul** [comment]
Thanks I’ll take a look!

(if you are just using MPE to retune by the way you know surge has full blown built in tuning including support for ODDSound?)

**Comment by nylonius** [comment]
i don't think ODDSound supports keyboards with more than 128 keys such as Lumatone (yet). When there are more notes than 128 in the tuning, MPE is necessary, unless one uses several instances of the synth on multiple channels, or the synth has a multichannel mode like Pianoteq and Surge XT do. (The latter is limited to tunings with the channels separated by octaves, though.)

**Comment by baconpaul** [comment]
this mid isn't valid MPE by the way. There are same channel note ons for notes which haven't been closed off. 
Is there any way to record what you actually send to surge?
I'll try and figure it out from this though.

**Comment by baconpaul** [comment]
but it would be really nice if i knew exactly what midi you were sending. Like: what channel is the actually poly at message coming on, with which key. is there no way you can capture that?

**Comment by baconpaul** [comment]
Alright so looking at this i don't think this is anything to do with MPE

If I just take a single note and send poly at on channel 5 with the note on channel 5, we get nothing

so I think the recent fix wasn't complete.

Here's the midi data I send in non MPE mode with poly at bound to pitch. On channel 1 fine on channel 5 no bend.

<img width="501" alt="Screen Shot 2022-02-24 at 2 03 36 PM" src="https://user-images.githubusercontent.com/13387561/155590221-9cda3f8d-4883-40cc-8921-69a6d0e40e88.png">

**Comment by baconpaul** [comment]
and yeah caught this in a regtest. Interesting. Stay tuned.

**Comment by cgibbard** [comment]
Ah, that's interesting! I haven't used MPE in conjunction with my Lumatone, but I've definitely played with the Poly AT stuff in multi-channel layouts and had it work, so maybe our recent change broke something else...

**Comment by baconpaul** [comment]
i got it working. it's nothing to do with MPE. 2 Character logic error in the multi channel poly at :)

**Comment by baconpaul** [comment]
Anyway that PR contains an assertive regtest to make sure poly at works on channels, and the tiny fix needed to make it pass.

@cgibbard the idiom `arr[chan & 16]` doesn't mask. You want `arr[chan & 15]` to mask to values < 16 :)

**Comment by baconpaul** [comment]
Oh and @nylonius this means a nightly in about 90 minutes or so should be fixed for you. Thanks for the report!

**Comment by cgibbard** [comment]
> Anyway that PR contains an assertive regtest to make sure poly at works on channels, and the tiny fix needed to make it pass.
> 
> @cgibbard the idiom `arr[chan & 16]` doesn't mask. You want `arr[chan & 15]` to mask to values < 16 :)

Oof, ouch! Yeah, as soon as I saw the diff, I knew what happened.

**Comment by baconpaul** [comment]
Ha! Well good bug reports make good software. Glad we found it!

**Comment by nylonius** [comment]
Sounds great! lookingforward, thanks so much. Not that it matters when this is not about MPE but I'm puzzled by what you said about the MIDI file I sent - it has midi notes with aftertouch on five midi channels, which I thought would be enough to do the test. Should you still need the MPE-tuned version form UTE, I can record that in Logic (Cubase has the bug that prevents it).

**Comment by baconpaul** [comment]
Oh the file you sent was just multi channel polyphonic midi which is different from valid mpe. But it still helped me find the bug so no need for any action

**Comment by nylonius** [comment]
RIght, that was because I couldn't record the MPE in Cubase but even when played just like that  AT worked in all those other synths but not Surge so I figured it can be used for testing. -Glad to hear this is going to get fixed, and so soon!

**Comment by baconpaul** [comment]
New nightly with the fix is up. If you grab it it should solve your problem. If not please reopen this with the details. Thanks again for the report!

**Comment by nylonius** [comment]
Hi. One of the options now works: the multichannel mapping in the  Tune mode. It tunes correctly and the synth now responds to poly AT. However, when MPE-tuned, Surge still doesn't react to poly AT. The plug-in version tunes the notes correctly, though, but the stand-alone version doesn't: it plays completely random-sounding notes. So again MPE disables poly AT, and the standalone version even responds errroneously to MPE tuning. Tomorrow I will record the multichannel-pitchbend-tuned MIDI from Lumatone>UTE> so that you can look at and test. Perhaps there's something in the MIDI from UTE that Surge doesn't accept, although it does work for other MPE synths. The MIDI for the multichannel tuning in Tune - which works - comes directly from Lumatone, not via UTE.  So maybe Surge doesn't like UTE.

**Comment by baconpaul** [comment]
So the only way I'll be able to debug that is if you as you say record what is coming out of the back of UTE. Or make it so I can run UTE. But to do that, I need to know what UTE is :)

Alternately you can just ship me a lumatone and I'm happy to debug. Chuckle.

The standalone and the plugin version *should* respond identically to midi but it occurs to me that in the standalone the routing is different. If you go options / audio midi settings is some sub-set of your input not selected in the midi section? That's the primary thing I would think of.

**Comment by baconpaul** [comment]
I also took surge AU in Logic pro
sent it the midi I screenshot above
and in MPE mode I get the poly at no problem

So I wonder if the UTE is sending the poly at on the global channel in mpe mode? but then it wouldn't work.

basically all the tests I can construct seem to work fine; so really need to be able to run or grab midi from the output of your system.

**Comment by mkruselj** [comment]
https://hpi.zentral.zone/ute

**Comment by nylonius** [comment]
The routing settings were identical in the standalone and in the plug-in, so that wasn't the reason. - UTE does send the poly in individual channels; it works correctly in several MPE synths. Will send mid file tomorrow.

**Comment by baconpaul** [comment]
Great! Like I said, when I send poly at on a channel matching a note in Logic Pro I get no problem. I don't have a way to route that to the standalone easily. (Thinking of it: I should make the standalone so you can drop a .mid file on it and it renders!)

**Comment by nylonius** [comment]
Problem solved! I should have realized this but even though Lumatone can send poly aftertouch, the aftertouch in pitchbend-tuned polyphonic MIDI from UTE is not poly but _channel_ aftertouch. The sounding result is the same, because each individual note gets dynamically thrown to its own MIDI channel. When channel aftertouch is chosen as the modulation source in this situation, it works.  The only thing that still causes problems is the MPE mode in the standalone version - wrong notes (the MPE pitchbend range settings are correct, so it's not that.) - I recorded a MIDI file that shows what the output from UTE looks like. There's probably no use for this any more but I'll send it anyway, having struggled all day with Logic's Environment and its confusing routing settings and finally succeeded to produce this
[UTE Aftertouch test.mid.zip](https://github.com/surge-synthesizer/surge/files/8141212/UTE.Aftertouch.test.mid.zip)
.

**Comment by baconpaul** [comment]
Yes as I said at the top, poly at isn’t valid mpe! Surge swaps channel at to a voice level modulator in mpe mode. (We still had a bug with multi channel non mpe poly at though which is what we fixed). I’m glad both Ute and surge handle mpe properly in that regard!

I will keep this open and look at the midi file. There’s really no difference between standalone and the plugin on midi handling thoigh so I’m can’t even think of what the problem would be. But an example helps! Thanks for sharing

**Comment by nylonius** [comment]
Well, here the situation is simple: in the standalone version, the MPE switch doesn't activate MPE. If I send the above midi file, for example, to the plug-in, there is a difference if MPE is active - that's the correct tuning and polyphonic behavior (with MPE pitchbend range at 2 semitones). In the standalone, with the exact same sound preset and settings, there is no difference, and what it plays is not the correct MPE version (with multiple channels).

**Comment by baconpaul** [comment]
I just plugged in my linnstrument (which sends MPE) turned the standalone into MPE mode and was able to use all gestures.

Here's my settings screen

<img width="514" alt="Screen Shot 2022-02-25 at 11 23 48 AM" src="https://user-images.githubusercontent.com/13387561/155750542-e7ee498a-55dd-44a5-8bad-ca5a094a33b3.png">

I really think something is wrong with the way you system is sending MIDI to the standalone. If I send a valid MPE stream and turn on MPE it works fine. And there is literally zero code difference between the standalone and au midi handling.

How do you make UTE send midi to the standalone exactly?

**Comment by baconpaul** [comment]
Hmm right and for completeness, I took your midi file dragged it into logic and set the plugin as external instrument on IAC; made the standalone listen on IAC and output to a loopback; and recorded that loopback back in logic (so basically used the standalone as an external synth in LPX) and no problem. As long as I route track 2 on track 2 track 3 on track 3 etc... then i got audio. 

gh doesn't let me attach an mp3 but here's a zipped mp3 of what i heard with a little patch where i modulate filter on channel at. my standalone is still set up for linn with mpe pitchbend range at 24.

[Standalone.mp3.zip](https://github.com/surge-synthesizer/surge/files/8143393/Standalone.mp3.zip)

**Comment by nylonius** [comment]
I'd already done the same test in Cubase but to make sure there's not some problem in the settings or external midi handling in Cubase, I tried playing that midi file in Logic, first sending the midi to the XT plug-in (in MPE mode) and then to the standalone with the exact same sound settings, via an IAC port. The result is completely different(notes all over the place in standalone). I also checked from a midi monitor app that the IAC port from Logic is indeed sending notes and pitchbends on multiple channels, and it is. Finally, I sent the midi via IAC to a standalone instance of Pianoteq, another synth that handles multichannel pitchbend-tuning, and the notes were correct. So if there's nothing wrong with the XT standalone module, I must have some settings wrong in it but what? Can't think of anything. The pb range is correct, changing the sample rate had no efffect.

**Comment by nylonius** [comment]
I found it! 'Use MIDI Channel for Octave Shift' was selected in the Tune window and that caused the wrong pitches. When unchecked, the MPE tuning comes out right. I didn't expect the settings in the Tune window to affect the tuning when MPE mode is on (and Tune mode off). Is this intentional or is it a bug?
I still can't get aftertouch working in the standalone version but the problem is not in Surge. I have four different standalone synths and _none_ of them respond to aftertouch, whether it's sent directly from a keyboard or from software; the plug-in versions do. I wonder if this is some kind of issue with the Mac or the OS?

**Comment by baconpaul** [comment]
Wonderful!

I wonder if grabbing a midi monitor and seeing if the aftertouch comes through IAC would be good. It sounds like something is making IAC not send AT for you. But when I sent the poly AT from logic -> standalone I was able to modulate with it in non-MPE mode.

So it is definitely the case that octave mode and MPE conficlit. I thought we had addressed that (so that octave mode is ignored in MPE mode) everywhere but we obviously missed at least one spot. I'll change the topic of this to reflect that and also tag @cgibbard  - one of he or i will fix it before the next release.

Thank you for your patience and the excellent bug report. Glad we identified the problem.

**Comment by nylonius** [comment]
Great! - I don't get standalone synths receive aftertouch even when the source is a midi keyboard (Lumatone) via USB rather than an IAC port.

**Comment by baconpaul** [comment]
wow ok so i should try that obviously! I think one of my controllers sends poly at.

**Comment by cgibbard** [comment]
The Lumatone's poly AT works for me in various standalone synths including Surge and PianoTeq, and I'm plugged in via USB, but I'm on Linux. I just tried it on main, and everything seemed to work, so I'm not sure what's up there.

I also tried checking the "MPE" flag at the top of the UI, and I didn't notice anything broken when in Channel per Octave mode... but then again, my Lumatone isn't producing MPE output, it's producing channel-per-octave 31edo MIDI. If it *were* producing MPE, then the channels would probably be getting mapped to octaves, and you'd probably have each note at a mostly-random octave (it would be the least recently used one if your controller is smart about how it assigns channels). So, we probably should force-disable the channel-per-octave setting when MPE is turned on.

**Comment by nylonius** [comment]
>If it _were_ producing MPE, then the channels would probably be getting mapped to octaves, and you'd probably have each note at a mostly-random octave 
Yes, exactly, that's what happens for MPE tuned notes when it's on. I think the functions under the Tune button should be bypassed when it is not active.

I continue to get slightly confusing results with AT in standalones. I did now get pianoteq respond to it when sent from Lumatone; I don't know what was wrong previously. It may be that I used a Lumatone mapping preset that was created in the earlier version of the .ltn editor, known to cause problems with AT. In any case, my situation is now this: aftertouch works but not in standalone instruments when using MPE. I can live with this.
